src/emulator/storage/files.ts

```diff
@@ -464,6 +477,88 @@ export class StorageLayer {
   public get dirPath(): string {
     return this._persistence.dirPath;
   }
+
+  /**
+   * Export is implemented using async operations so that it does not block
+   * the hub when invoked.
+   */
+  async export(storageExportPath: string) {
+    // Export a list of all known bucket IDs, which can be used to reconstruct
+    // the bucket metadata.
+    const bucketsList: BucketsList = {
+      buckets: [],
+    };
+    for (const b of this.listBuckets()) {
+      bucketsList.buckets.push({ id: b.id });
+    }
+    const bucketsFilePath = path.join(storageExportPath, "buckets.json");
+    await fse.writeFile(bucketsFilePath, JSON.stringify(bucketsList, undefined, 2));
+
+    // Recursively copy all file blobs
+    const blobsDirPath = path.join(storageExportPath, "blobs");
+    await fse.ensureDir(blobsDirPath);
+    await fse.copy(this.dirPath, blobsDirPath, { recursive: true });
+
+    // Store a metadata file for each file
+    const metadataDirPath = path.join(storageExportPath, "metadata");
+    await fse.ensureDir(metadataDirPath);
+
+    for await (const [p, file] of this._files.entries()) {
+      const metadataExportPath = path.join(metadataDirPath, p) + ".json";
+      const metadataExportDirPath = path.dirname(metadataExportPath);
+
+      await fse.ensureDir(metadataExportDirPath);
+      await fse.writeFile(metadataExportPath, StoredFileMetadata.toJSON(file.metadata));
+    }
+  }
+
+  /**
+   * Import can be implemented using sync operations because the emulator should
+   * not be handling any other requests during import.
+   */
+  import(storageExportPath: string) {
+    // Restore list of buckets
+    const bucketsFile = path.join(storageExportPath, "buckets.json");
+    const bucketsList = JSON.parse(fs.readFileSync(bucketsFile).toString()) as BucketsList;
+    for (const b of bucketsList.buckets) {
+      const bucketMetadata = new CloudStorageBucketMetadata(b.id);
+      this._buckets.set(b.id, bucketMetadata);
+    }
+
+    // Restore all metadata
+    const metadataDir = path.join(storageExportPath, "metadata");
+    const metadataList = this.walkDirSync(metadataDir);
+    for (const f of metadataList) {
+      const metadata = StoredFileMetadata.fromJSON(
+        fs.readFileSync(f).toString(),
+        this._cloudFunctions
+      );
+
+      const metadataPath = path.relative(metadataDir, f);
+      const blobPath = metadataPath.substring(0, metadataPath.length - 5); // Subtract .json
+
+      const file = new StoredFile(metadata, blobPath);
+      this._files.set(blobPath, file);
+    }
+
+    // Recursively copy all blobs
+    const blobsDir = path.join(storageExportPath, "blobs");
+    fse.copySync(blobsDir, this.dirPath);
+  }
+
+  private walkDirSync(dir: string, res: string[] = []): string[] {
```

Comment:
Optional: TBH this would be more readable if rewritten as a generator function with `yield` and `yield*`

Reply:
Learning new things!